hp-Adaptivity: through local smoothness & prior error estimate

skvarjun opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

**Pure h adaptivity may leads to poor convergence in certain scanarios;**

1. Singularities and sharp gradients (e.g., crack tips, re-entrant corners as in 'the L-shaped domains'): require local h-refinement and p-enrichment to accurately capture stress concentrations.
2. Boundary layers and multi-scale phenomena: demand h-refinement and variable polynomial orders to resolve steep gradients without global mesh refinement.
3. Solution smoothnes varying across the domain: hp-adaptivity adapts  higher polynomial degree in smooth regions and mesh size in non-smooth zones for efficiency.


## Design
<!--A concise description (design) of the enhancement.-->
The key concept used here is to compute the solution smoothness in elements. Here the finite solution is projected onto Legendre polynomial and then the Legendre coeffient decay rate is used to estimate the smoothness; faster the decay, smoother the solution be. 

**Detailed hp-Adaptivity algorithm : logic flow**
1. Projecting local solution in each element to Legendre expansion.
2. Added a function legepoly() to compute tensor-product Legendre basis functions evaluated at a point p.
3. Use the classical recurrence relation to evaluate 1D Legendre polynomials L_n(x), extended to multiple dimensions.
4. Construct the basis up to a specified total degree order of the element.
5. For each element, extract the local dof and compute the finite element solution u(h) at quadrature points.
6. Assemble the projection matrix K and right-hand side  F using   K_ij = integral (psi_i psi_i) dx and F_i = u_h psi_i dx
7. Solve the system K.c=F to obtain the Legendre coefficients c per element.
8. Then index-to-Degree Mapping: Here, a lookup vector mapping each index in the coefficient vector c to its corresponding total polynomial degree.
9. This mapping gives the basis ordering used during legepoly() construction
10. Coefficient Grouping and Filtering: Group coefficients c_i by their associated total degree.
11. From |k_i|, omitting all coefficients where ∣k∣=0 to exclude the constant mode. Then accumulate all coefficients (U_k) of the same degree into vectors keyed by degree.
12. Then for each degree (k >= 1), compute the L2 norm.
13. Finally, log-Log regression for smoothness esimation.
14. Construct vectors log(∣U_k∣) and log(k) using filtered norms.
15. Apply least-squares linear regression as log(∣U_k∣) = −s.log(k) + C
16. Compute the negtaive of slope s (call it regularity), and define the smoothness index as for the given element.
17. Finally calculates p corsen (`p_error`) and h corsen error (`h_error`) (as @roystgnr composed in hp_coarsentest.C) and then if the element's local solution is having highest regularity than mean (this can be tuned by user - haven't added yet, may be later) and `p_error > h_error` then do p refine else h refine.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
1. Computational cost optimization: combining h- and p-adaptivity reduces degrees of freedom while preserving accuracy, especially in large-scale or real-time simulations.
2. Faster convergency with the optimal DoF in the system. 

Input sample:  
[Executioner]
  [Adaptivity]
    hp_refinement = true
    steps = 6
    refine_fraction = 0.7
    coarsen_fraction = 0.05
    max_h_level = 15
  []
[]


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

skvarjun committed:
hp-Adaptivity: through smoothness and h/p refine errors

lindsayad commented on framework/include/base/Adaptivity.h:
```diff
@@ -313,6 +316,8 @@ class Adaptivity : public ConsoleStreamInterface,
   std::map<std::string, std::unique_ptr<libMesh::ErrorVector>> _indicator_field_to_error_vector;
 
   bool _p_refinement_flag = false;
+
+  bool _hp_refinement_flag = false;
```
I don't know if we have an enum like that already in libMesh @roystgnr ?

skvarjun replied:
Added. 
```ruby
MooseEnum adaptivity("h p hp", "h");
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3082728/), step Clang format on 5363306 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31109/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31109/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2ad78df8c544c9cf2cf0b743b06b5557ce23a663`

skvarjun committed:
hp-Adaptivity: enum for mesh adaptivity - adaptivity_type h or p or hp
refs #31109

skvarjun commented:
Some tests may fail as I added a single moose enum feature for switching between h, p and hp adaptivity (default is h). So, those tests may have to be reformatted.

lindsayad commented on framework/include/base/Adaptivity.h:
```diff
@@ -300,6 +302,9 @@ class Adaptivity : public ConsoleStreamInterface,
   /// Name of the marker variable if using the new adaptivity system
   std::string _marker_variable_name;
 
+  /// Type of mesh adaptivity
+  std::string _adaptivity_type;
```
```suggestion
  enum class AdaptivityType
  {
    H = 0,
    P = 1,
    HP = 2
  };

  /// Type of mesh adaptivity
  AdaptivityType _adaptivity_type;
```

lindsayad replied:
The enum class definition will need to be moved earlier in the class, before the `init` declaration

lindsayad replied:
You could also move it outside the `Adaptivity` class definition if you want to avoid constant `Adaptivity::` scoping

lindsayad commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -40,6 +40,8 @@ AdaptivityAction::validParams()
   MooseEnum estimators("KellyErrorEstimator LaplacianErrorEstimator PatchRecoveryErrorEstimator",
                        "KellyErrorEstimator");
 
+  MooseEnum adaptivity("h p hp", "h");
+
```
```suggestion
```

lindsayad commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -52,6 +54,8 @@ AdaptivityAction::validParams()
                         "The fraction of elements or error to coarsen. Should be between 0 and 1.");
   params.addParam<MooseEnum>(
       "error_estimator", estimators, "The class name of the error estimator you want to use.");
+  params.addParam<MooseEnum>(
+      "adaptivity_type", adaptivity, "The type of the mesh adaptivity you want to use.");
```
```suggestion
```

lindsayad commented on framework/src/actions/SetAdaptivityOptionsAction.C:
```diff
@@ -43,7 +43,9 @@ commonAdaptivityParams()
   params.addParam<bool>(
       "recompute_markers_during_cycles", false, "Recompute markers during adaptivity cycles");
-  params.addParam<bool>("switch_h_to_p_refinement", false, "True to perform p-refinement");
+  MooseEnum adaptivity("h p hp", "h");
```
```suggestion
  MooseEnum adaptivity("h=0 p=1 hp=2", "h");
```

lindsayad commented on framework/include/base/Adaptivity.h:
```diff
@@ -61,7 +61,9 @@ class Adaptivity : public ConsoleStreamInterface,
-  void init(const unsigned int steps, const unsigned int initial_steps, const bool p_refinement);
+  void init(const unsigned int steps,
+            const unsigned int initial_steps,
+            const MooseEnum & adaptivity_type);
```
```suggestion
            const AdaptivityType adaptivity_type);
```

lindsayad commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -140,11 +144,11 @@ AdaptivityAction::act()
     // can't just skip it for the useSplit case.
     if (_mesh->isSplit())
-      adapt.init(0, 0, getParam<bool>("switch_h_to_p_refinement"));
+      adapt.init(0, 0, getParam<MooseEnum>("adaptivity_type"));
```
```suggestion
      adapt.init(0, 0, getParam<MooseEnum>("adaptivity_type").getEnum<Adaptivity::AdaptivityType>());
```

lindsayad commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -140,11 +144,11 @@ AdaptivityAction::act()
       adapt.init(getParam<unsigned int>("steps"),
                  getParam<unsigned int>("initial_adaptivity"),
-                 getParam<bool>("switch_h_to_p_refinement"));
+                 getParam<MooseEnum>("adaptivity_type"));
```
```suggestion
                 getParam<MooseEnum>("adaptivity_type").getEnum<Adaptivity::AdaptivityType>());
```

lindsayad commented on framework/src/actions/SetAdaptivityOptionsAction.C:
```diff
@@ -144,7 +146,7 @@ SetAdaptivityOptionsAction::act()
     adapt.init(getParam<unsigned int>("steps"),
                getParam<unsigned int>("initial_steps"),
-               getParam<bool>("switch_h_to_p_refinement"));
+               getParam<MooseEnum>("adaptivity_type"));
```
```suggestion
               getParam<MooseEnum>("adaptivity_type").getEnum<Adaptivity::AdaptivityType());
```

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -58,7 +59,7 @@ Adaptivity::~Adaptivity() {}
 Adaptivity::init(const unsigned int steps,
                  const unsigned int initial_steps,
-                 const bool p_refinement)
+                 const MooseEnum & adaptivity_type)
```
```suggestion
                 const AdaptivityType adaptivity_type)
```

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -73,10 +74,20 @@ Adaptivity::init(const unsigned int steps,
   _initial_steps = initial_steps;
   _steps = steps;
-  _p_refinement_flag = p_refinement;
+
+  if (adaptivity_type == "h")
+    _adaptivity_type = std::string("h");
+  else if (adaptivity_type == "p")
+    _adaptivity_type = std::string("p");
+  else if (adaptivity_type == "hp")
+    _adaptivity_type = std::string("hp");
+  else
+    mooseError(std::string("Unknown mesh adaptivity type selection: ") +
+               std::string(adaptivity_type));
```
```suggestion
  _adaptivity_type = adaptivity_type;
```

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -73,10 +74,20 @@ Adaptivity::init(const unsigned int steps,
   _mesh_refinement_on = true;
 
-  if (_p_refinement_flag)
+  if (_adaptivity_type == "p" || _adaptivity_type == "hp")
```
```suggestion
  if (_adaptivity_type == AdaptivityType::P || _adaptivity_type == AdaptivityType::HP)
```

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -209,6 +220,12 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
+    // Moving some of h flagged elements to p flagged based on the
+    // local smoothness and prior h & p error estimates
+    HPCoarsenTest hpselector;
+    if (_adaptivity_type == "hp")
```
```suggestion
    if (_adaptivity_type == AdaptivityType::HP)
```

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -229,7 +246,7 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
     _mesh_refinement->make_flags_parallel_consistent();
 
-  if (_p_refinement_flag)
+  if (_adaptivity_type == "p")
```
```suggestion
  if (_adaptivity_type == AdaptivityType::P)
```

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -242,7 +259,7 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
       _displaced_mesh_refinement->make_flags_parallel_consistent();
 
-    if (_p_refinement_flag)
+    if (_adaptivity_type == "p")
```
```suggestion
    if (_adaptivity_type == AdaptivityType::P)
```

lindsayad commented on modules/level_set/src/transfers/LevelSetMeshRefinementTransfer.C:
```diff
@@ -55,7 +55,8 @@ LevelSetMeshRefinementTransfer::initialSetup()
       adapt.setCyclesPerStep(from_problem.adaptivity().getCyclesPerStep());
-      adapt.init(1, 0, false);
+      MooseEnum adaptivity_type("h p hp", "h");
+      adapt.init(1, 0, adaptivity_type);
```
```suggestion
      adapt.init(1, 0, adaptivity_type.getEnum<Adaptivity::AdaptivityType>());
```

skvarjun committed:
hp-Adaptivity: adaptivity enumaration scheme updated

lindsayad reviewed (changes requested):
We need an hp-refinement test

skvarjun commented on framework/src/base/Adaptivity.C:
```diff
@@ -209,6 +212,12 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
 
+    // Moving some of h flagged elements to p flagged based on the
+    // local smoothness and prior h & p error estimates
+    HPCoarsenTest hpselector;
```
Yes, it is. I remember @roystgnr mentioned that `HPCoarsenTest` was originally a test to see how efficient the hp selection algorithm was.I suppose changing the class name to somthing like `HPAdaptivity` would be nice.

skvarjun commented on framework/src/base/Adaptivity.C:
```diff
@@ -229,7 +238,7 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
+  if (_adaptivity_type == AdaptivityType::P)
     _mesh_refinement->switch_h_to_p_refinement();
 
   // Perform refinement and coarsening
```
In my opinion, it'd be nice if `set_refinement_flag` shall be `set_h_refinement_flag` to name h refinement

skvarjun committed:
hp-Adaptivity: test for hp adaptivity added

skvarjun committed:
hp-Adaptivity: added SiblingCoupling functor for parallel

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3138821/), step Clang format on 51ebc4b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31109/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31109/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9fc4b031c22bae4a57a91dc6fdc49bed315b38f5`

skvarjun committed:
hp-Adaptivity: updated gold test file added

skvarjun committed:
hp-Adaptivity:  deleted unused gold file

skvarjun committed:
hp-Adaptivity: deleted outdated gold file

skvarjun committed:
hp-Adaptivity: new gold file for hp adaptivity

skvarjun commented:
@lindsayad Could you have a look at this PR?

lindsayad committed:
Tweak doxygen for new Adaptivity data members

lindsayad committed:
Initialize _adaptivity_type to H

lindsayad committed:
Add sibling coupling to all DofMaps

lindsayad committed:
Remove captain obvious comment and simplify if (_displaced_problem) conditional

lindsayad committed:
Revert "Add sibling coupling to all DofMaps"

This reverts commit e7e72f1f47e25f49f4285978d37d68f0f5fe90db.

We currently only will inspect the solution vector in the zeroth
nonlinear system when determining whether we get more error
reduction by p-refining than h-refining.

lindsayad committed:
Only create SiblingCoupling if doing HP refinement

lindsayad committed:
Remove AdaptivityAux

lindsayad committed:
Some Adaptivity generalizations

- Instead of always pulling the zeroth nonlinear system, pull the
  zeroth solver system
  - In conjunction with this change, error out if we have more than
    one solver system
- Perform hp flag toggling past the if-else branch for new vs. old
  adaptivity system

lindsayad committed:
Add hp-adaptivity test for new adaptivity system

lindsayad committed:
Change shouldComputeInternalSide to only query the elem and neighbor

There's no need to go to a MooseMesh API because there's no need to
enforce a choice based off one global rule when we may make decisions
purely using local info

lindsayad committed:
Add MooseMesh APIs for whether we're doing HP-refinement

We must query these in ThreadedElementLoopBase to check whether
we might possibly doing h-refinement... or do we...

lindsayad committed:
Use MooseMesh APIs in shouldComputeInternalSide

lindsayad committed:
Mostly revert "Add MooseMesh APIs for whether we're doing HP-refinement"

This mostly reverts commit 6e040c6c8ac97a19668a2d6ba2ef182a4d3c2d0b.

The only changes we keep are the changes to the doxygen for the MooseMesh
p-refinement related members

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3456196/), step Results summary on cd857cd wanted to post the following:

## Framework test summary

Compared against 65c471a in job [civet.inl.gov/job/3452904](https://civet.inl.gov/job/3452904).

### Added tests

| Test                                    | Time (s)   | Memory (MB)   |
|-----------------------------------------|------------|---------------|
| `adaptivity/hp_adaptivity.test/user`    | 1.67       | 138.69        |
| `adaptivity/hp_adaptivity.test/libmesh` | 1.33       | 124.53        |

### Run time changes

| Test                                                                                               | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|----------------------------------------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `outputs/console/multiapp.ts_both_multi_output`                                                    | 2.99       | 5.55       | +85.72% | 134.17      | 147.39      |
| `multiapps/secant.variables_transient/app_begin_transfers_end`                                     | 4.88       | 8.56       | +75.48% | 139.31      | 121.80      |
| `multiapps/secant.variables_transient/sub_cycling`                                                 | 2.19       | 3.72       | +69.95% | 122.61      | 119.59      |
| `multiapps/steffensen.variables_transient/app_begin_transfers_begin`                               | 2.22       | 3.73       | +68.10% | 119.18      | 129.33      |
| `multiapps/steffensen.variables_transient/sub_cycling`                                             | 2.03       | 3.39       | +66.84% | 132.85      | 123.23      |
| `multiapps/picard.rel_tol`                                                                         | 3.90       | 6.49       | +66.35% | 139.56      | 126.67      |
| `multiapps/secant_postprocessor.pp_transient/iteration_adaptive`                                   | 2.37       | 3.91       | +64.53% | 126.95      | 117.52      |
| `time_integrators/convergence.implicit_astabledirk4/level1`                                        | 3.10       | 5.06       | +63.10% | 135.42      | 132.84      |
| `multiapps/picard.test`                                                                            | 4.17       | 6.80       | +62.92% | 120.12      | 133.10      |
| `outputs/console/multiapp.ts_end_multi_output`                                                     | 2.20       | 3.54       | +60.70% | 121.70      | 112.67      |
| `executioners/eigen_executioners.test_inverse_power_method`                                        | 2.72       | 4.37       | +60.47% | 126.65      | 169.04      |
| `multiapps/secant.variables_transient/iteration_adaptive`                                          | 2.59       | 4.14       | +60.01% | 115.23      | 121.36      |
| `multiapps/picard.iteration_adaptive`                                                              | 4.60       | 7.36       | +59.81% | 131.64      | 148.94      |
| `multiapps/secant.variables_transient/app_begin_transfers_begin`                                   | 2.47       | 3.95       | +59.70% | 120.38      | 127.34      |
| `multiapps/picard_postprocessor.pp_transient/app_begin_transfers_end`                              | 4.74       | 7.54       | +59.25% | 161.79      | 134.42      |
| `executioners/multisys-ss-detection.multi-sys-ss`                                                  | 6.36       | 10.11      | +58.86% | 112.75      | 140.27      |
| `multiapps/steffensen.variables_transient/app_begin_transfers_begin_steffensen_sub`                | 2.39       | 3.76       | +57.12% | 155.32      | 117.17      |
| `multiapps/steffensen.variables_transient/iteration_adaptive`                                      | 2.51       | 3.93       | +56.76% | 114.91      | 114.89      |
| `multiapps/picard_postprocessor.pp_transient/app_begin_transfers_begin`                            | 4.04       | 6.30       | +56.00% | 122.18      | 129.56      |
| `transfers/multiapp_postprocessor_interpolation_transfer.basis/from_sub_to_parent_single_sub_file` | 2.37       | 3.69       | +55.26% | 153.01      | 144.36      |
| `multiapps/picard_postprocessor.pp_transient/sub_cycling`                                          | 3.79       | 5.88       | +55.22% | 131.31      | 132.99      |
| `time_integrators/convergence.implicit_astabledirk4/level2`                                        | 5.00       | 7.74       | +54.93% | 129.54      | 143.50      |
| `kernels/ode.test_parsed_sys_impl`                                                                 | 2.17       | 3.34       | +53.91% | 112.39      | 147.83      |
| `outputs/console/multiapp.ts_begin_multi_output`                                                   | 2.50       | 3.85       | +53.76% | 116.44      | 119.83      |
| `transfers/transfer_interpolation.test`                                                            | 2.88       | 4.42       | +53.65% | 152.37      | 133.83      |
| `multiapps/secant_postprocessor.pp_transient/test_rerun`                                           | 2.49       | 3.80       | +52.68% | 119.01      | 113.58      |
| `multiapps/steffensen.variables_transient/app_end_transfers_end`                                   | 2.06       | 3.14       | +52.42% | 119.48      | 124.15      |
| `multiapps/picard_postprocessor.pp_transient/iteration_adaptive`                                   | 4.19       | 6.35       | +51.42% | 119.56      | 133.09      |
| `multiapps/relaxation.standard`                                                                    | 2.07       | 3.12       | +50.50% | 139.51      | 130.27      |
| `make_install.test_copy_install/run`                                                               | 3.62       | 5.45       | +50.33% | 254.87      | 266.35      |
## Modules test summary

Compared against 65c471a in job [civet.inl.gov/job/3452904](https://civet.inl.gov/job/3452904).

### No added tests


### Run time changes

| Test                                                                                                                    | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|-------------------------------------------------------------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `thermal_hydraulics/test:problems/water_hammer.3eqn`                                                                    | 2.73       | 5.17       | +89.49% | 140.55      | 130.60      |
| `thermal_hydraulics/test:components/gate_valve_1phase.1phase_light`                                                     | 2.91       | 5.38       | +84.87% | 172.97      | 167.54      |
| `thermal_hydraulics/test:problems/pressure_drop.tests/without_junction`                                                 | 2.73       | 4.80       | +75.62% | 124.63      | 139.98      |
| `stochastic_tools/test:transfers/sampler_reporter.transfer/normal`                                                      | 8.31       | 14.13      | +70.02% | 156.90      | 151.41      |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated-transient.channel_ERCOFTAC_linear_transient` | 7.21       | 12.26      | +70.00% | 139.47      | 143.80      |
| `solid_mechanics/test:combined_creep_plasticity.combined`                                                               | 6.67       | 11.26      | +68.70% | 118.99      | 126.62      |
| `solid_mechanics/test:rate_independent_cyclic_hardening.linear_kinharden_nonsymmetric_strain_controlled`                | 11.93      | 19.65      | +64.76% | 128.10      | 133.13      |
| `stochastic_tools/test:web_server_control.stochastic_control/normal`                                                    | 3.04       | 5.01       | +64.58% | 267.32      | 0.00        |
| `solid_mechanics/test:combined_creep_plasticity.creepWithPlasticity`                                                    | 6.24       | 10.27      | +64.55% | 116.77      | 110.52      |
| `solid_mechanics/test:combined_creep_plasticity.combined_start_time`                                                    | 6.84       | 11.18      | +63.36% | 120.03      | 110.68      |
| `solid_mechanics/test:dynamics/acceleration_bc.acceleration_bc`                                                         | 4.15       | 6.78       | +63.30% | 130.72      | 134.45      |
| `solid_mechanics/test:temperature_dependent_hardening.test`                                                             | 2.70       | 4.41       | +63.22% | 113.01      | 115.85      |
| `navier_stokes/test:finite_volume/ins/turbulence/block-restricted.channel_ERCOFTAC_linear_block`                        | 3.36       | 5.46       | +62.84% | 129.27      | 134.08      |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/segregated.channel_ERCOFTAC`                                   | 8.68       | 13.97      | +60.97% | 124.84      | 131.77      |
| `combined/test:adaptive_timestepping.test_function_change`                                                              | 4.31       | 6.89       | +59.92% | 158.63      | 165.15      |
| `thermal_hydraulics/test:problems/brayton_cycle.open`                                                                   | 2.05       | 3.26       | +59.35% | 175.76      | 186.44      |
| `solid_mechanics/test:combined_creep_plasticity.stress_prescribed`                                                      | 4.24       | 6.64       | +56.65% | 109.96      | 109.54      |
| `optimization/test:optimizationreporter/nonlinear_material.nonlinear_material`                                          | 2.79       | 4.31       | +54.59% | 267.02      | 266.93      |
| `contact/test:avoid-aligned-node-exception.test`                                                                        | 3.02       | 4.66       | +54.24% | 154.81      | 166.46      |
| `solid_mechanics/test:crystal_plasticity/cp_eigenstrains.thermal_eigenstrain`                                           | 4.94       | 7.53       | +52.41% | 119.61      | 119.21      |
| `contact/test:glued.glued_contact_mechanical_constraint`                                                                | 2.88       | 4.36       | +51.24% | 120.44      | 119.06      |
| `combined/test:combined_plasticity_temperature.temp_dep_yield`                                                          | 2.50       | 3.78       | +50.98% | 146.81      | 155.21      |
| `navier_stokes/test:finite_volume/pins/block-restriction/segregated.porous_block_restricted_simple`                     | 2.68       | 4.05       | +50.89% | 115.14      | 116.51      |
| `solid_mechanics/test:lagrangian/cartesian/total/rates.truesdell_shear`                                                 | 2.67       | 4.03       | +50.67% | 140.38      | 140.39      |
| `navier_stokes/test:finite_volume/cns/heated-channel.kt-van-leer-primitive`                                             | 2.22       | 3.34       | +50.62% | 120.52      | 117.83      |
| `solid_mechanics/test:material_limit_time_step/creep.test5a_lim`                                                        | 2.32       | 3.49       | +50.42% | 121.08      | 125.65      |
| `phase_field/test:grain_tracker_test.test_elemental`                                                                    | 6.12       | 2.14       | -64.98% | 173.28      | 128.04      |

GiudGiud reviewed (commented):
looks ready to me once it passes tests

lindsayad committed:
Don't compute internal sides for CacheChangedListsThread

This avoids our asserts in the base class implementation. Also there's
nothing to do for internal sides for this type of threaded loop

lindsayad committed:
Smoother wording for adaptivity_type doc string

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
@roystgnr also feel free to take a pass

roystgnr commented on framework/src/base/Adaptivity.C:
```diff
@@ -214,6 +232,22 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
+                 "(for C0 elements), and hessian (for C0 and C1 elements) differences between "
+                 "coarse and fine states, and the gradient and hessian will be functions of mesh "
+                 "displacements. Consequently it's possible that refinement flags may end up out "
+                 "of sync between reference and displaced meshes which would be disastrous");
```
Wait, surely we're not just throwing the same error estimators at each mesh independently for h or p refinement?  Do we not do our refinement on a single mesh and then copy flags over to the other?

roystgnr commented on modules/doc/content/newsletter/2023/2023_07.md:
```diff
@@ -6,8 +6,8 @@
-can simply set `switch_h_to_p_refinement=true` in the `Adaptivity` block.
-As suggested by the parameter name, only one of h- or p-refinement may be used
+can simply set `adaptivity_type=p` in the `Adaptivity` block.
+As suggested by the parameter name, only one of h- or p- or hp-refinement may be used
```
Oh, heck no.  This is the newsletter, not the manual.  We should be rewriting user documentation, not history.

lindsayad replied:
Hahaha will fix

roystgnr reviewed (changes requested):
Normally I'd be hostile to changing the user interface so drastically with no deprecation period, even when it's a change for the better like this, but IMHO p adaptivity users are probably all the sort of people who can cope.

We definitely need better documentation of the change than a Minitrue stealth edit to a two-year-old newsletter, though.

lindsayad committed:
Just sync flags from reference to displaced mesh

lindsayad committed:
Add documentation

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3459809/), step Results summary on ba44f8f wanted to post the following:

## Framework test summary

Compared against 73eede6 in job [civet.inl.gov/job/3459230](https://civet.inl.gov/job/3459230).

### Added tests

| Test                                    | Time (s)   | Memory (MB)   |
|-----------------------------------------|------------|---------------|
| `adaptivity/hp_adaptivity.test/libmesh` | 1.40       | 111.43        |
| `adaptivity/hp_adaptivity.test/user`    |            | 0.00          |
## Modules test summary

Compared against 73eede6 in job [civet.inl.gov/job/3459230](https://civet.inl.gov/job/3459230).

### No added tests


### Run time changes

| Test                                                    | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|---------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `phase_field/test:grain_tracker_test.test_elemental`    | 6.94       | 3.40       | -51.01% | 163.70      | 128.09      |
| `phase_field/test:grain_tracker_test.test_faux_element` | 5.75       | 2.42       | -57.84% | 320.82      | 293.89      |

lindsayad committed:
Only sync elements to displaced mesh when we have one, smh

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3461267/), step Results summary on a1855b8 wanted to post the following:

## Framework test summary

Compared against 73eede6 in job [civet.inl.gov/job/3459230](https://civet.inl.gov/job/3459230).

### Added tests

| Test                                    | Time (s)   | Memory (MB)   |
|-----------------------------------------|------------|---------------|
| `adaptivity/hp_adaptivity.test/user`    | 1.57       | 115.14        |
| `adaptivity/hp_adaptivity.test/libmesh` | 1.29       | 115.87        |
## Modules test summary

Compared against 73eede6 in job [civet.inl.gov/job/3459230](https://civet.inl.gov/job/3459230).

No change
